services/motion/builtin/testutils.go

```diff
@@ -0,0 +1,542 @@
+) base.Base {
+	t.Helper()
+
+	reg, ok := resource.LookupRegistration(base.API, wheeled.Model)
+	test.That(t, ok, test.ShouldBeTrue)
+
+	wheeledConf := &wheeled.Config{
+		WidthMM:              200,
+		WheelCircumferenceMM: 2000,
+		Left:                 []string{leftMotorName},
+		Right:                []string{rightMotorName},
+	}
+
+	baseConf := resource.Config{
+		Name:                baseName,
+		API:                 base.API,
+		Frame:               &referenceframe.LinkConfig{Geometry: &spatialmath.GeometryConfig{R: geomSize}},
+		ConvertedAttributes: wheeledConf,
+	}
+	deps := resource.Dependencies{}
+	deps[motors[leftMotorName].Name()] = motors[leftMotorName]
+	deps[motors[rightMotorName].Name()] = motors[rightMotorName]
+
+	wBase, err := reg.Constructor(ctx, deps, baseConf, logger)
+	test.That(t, err, test.ShouldBeNil)
+	return wBase.(base.Base)
+}
+
+func createEncoders(t *testing.T, ctx context.Context, logger logging.Logger) map[string]encoder.Encoder {
+	t.Helper()
+	leftconf := resource.Config{
+		Name:                leftMotorName + "_encoder",
+		API:                 encoder.API,
+		ConvertedAttributes: &fakeencoder.Config{UpdateRate: int64(updateRate)},
+	}
+	rightconf := resource.Config{
+		Name:                rightMotorName + "_encoder",
+		API:                 encoder.API,
+		ConvertedAttributes: &fakeencoder.Config{UpdateRate: int64(updateRate)},
+	}
+	left, err := fakeencoder.NewEncoder(ctx, leftconf, logger)
+	test.That(t, err, test.ShouldBeNil)
+	right, err := fakeencoder.NewEncoder(ctx, rightconf, logger)
+	test.That(t, err, test.ShouldBeNil)
+	return map[string]encoder.Encoder{leftMotorName: left, rightMotorName: right}
+}
+
+func createMotors(
+	t *testing.T,
+	ctx context.Context,
+	logger logging.Logger,
+	encoders map[string]encoder.Encoder,
+) map[string]motor.Motor {
+	t.Helper()
+	motorMap := map[string]motor.Motor{}
+	for name, encoder := range encoders {
+		conf := resource.Config{
+			Name:                name,
+			API:                 motor.API,
+			ConvertedAttributes: &fakemotor.Config{Encoder: name + "_encoder", TicksPerRotation: 4000},
+		}
+
+		thisMotor, err := fakemotor.NewMotor(
+			ctx,
+			resource.Dependencies{encoder.Name(): encoder},
+			conf,
+			logger,
+		)
+		test.That(t, err, test.ShouldBeNil)
+		motorMap[name] = thisMotor
+	}
+	return motorMap
+}
+
+type noisyMovementSensor struct {
+	movementsensor.MovementSensor
+	mu         sync.Mutex
+	noise      spatialmath.Pose
+	queryCount int // Apply noise every other iteration when this is > 1
+}
+
+func (m *noisyMovementSensor) Position(ctx context.Context, extra map[string]interface{}) (*geo.Point, float64, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.queryCount++
+	if m.queryCount <= 10 || m.queryCount%2 == 0 {
+		return m.MovementSensor.Position(ctx, extra)
+	}
+	pos, alt, err := m.MovementSensor.Position(ctx, extra)
+	if err != nil {
+		return nil, 0, err
+	}
+	newPos := spatialmath.PoseToGeoPose(spatialmath.NewGeoPose(pos, 0), m.noise)
```

Comment:
why is heading set to zero, this doesn't seem like it'll always be the case

Reply:
should't we call compassheading within this function to get the actual heading value?

Reply:
The `Position` function here does not return a heading, just a lat/lon and an altitude. As the heading at the geopose does not affect the location of the geopose, it is not useful information and can always be zero.